### PR TITLE
[v6r20] VOMS2CS synchronization using REST VOMS service interface

### PR DESCRIPTION
**Pull request by atsareg**
The PR contains reimplementation of the VOMSService client to use the REST interface of the VOMS service, which is much simpler to use (no SOAP) and provides more comprehensive information. Iy allowed, in particular, not to import Suspended users from the VOMS server. Also added a new command  dirac-admin-voms-sync useful to run single-time synchronization and to examine user modifications for a given VO. 

BEGINRELEASENOTES
*Core
NEW: dirac-admin-voms-sync - command line for VOMS to CS synchronization
NEW: VOMS2CSSynchronizer - new class encapsulating VOMS to CS synchronization
CHANGE: VOMSService - reimplemented using VOMS REST interface
CHANGE: VOMS2CSAgent - reimplemented to use VOMS2CSSynchronizer
ENDRELEASENOTES

### CODE REVIEW

File: ConfigurationSystem/Client/VOMS2CSSyncronizer.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    self.vomsVOName = getVOOption(vo, "VOMSName", "")
+    if not self.vomsVOName:
+      raise Exception("VOMS name not defined for VO %s" % vo)
+    self.adminMsgs = {'Errors': [], 'Info': []}
+    self.vomsUserDict = {}
+    self.autoModifyUsers = autoModifyUsers
+    self.autoAddUsers = autoAddUsers
+    self.autoDeleteUsers = autoDeleteUsers
+
+  def syncCSWithVOMS(self):
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSyncronizer.py]: This method is super long, and difficult to read. Can it be split in few functions? And then unit-tested.

File: ConfigurationSystem/Client/VOMS2CSSyncronizer.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+
+    resultDict['CSAPI'] = self.csapi
+    resultDict['AdminMessages'] = self.adminMsgs
+    return S_OK(resultDict)
+
+  def getVOUserData(self, refreshFlag=False):
+    """ Get a report for users of a given VO
+
+    :param str vo: VO name
+    :return: S_OK/S_ERROR, Value = user description dictionary
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSyncronizer.py]: Not the correct docstring

File: ConfigurationSystem/Client/VOMS2CSSyncronizer.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    result = self.csapi.describeUsers(diracUsers)
+    if not result['OK']:
+      self.log.error('Could not retrieve CS User description')
+    return result
+
+  def getVOUserReport(self):
+    """
+
+    :param str vo: VO name
+    :return: report string
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSyncronizer.py]: Not the correct docstring

**Comment by fstagni** [comment]
There's no unit test. 

I think you also need to update the documentation in 
docs/source/AdministratorGuide/Configuration/ConfReference/Systems/Configuration/Agents/VOMS2CSAgent/index.rst

File: ConfigurationSystem/scripts/dirac-admin-voms-sync.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+for msg in resultDict["AdminMessages"]["Info"]:
+  gLogger.notice(msg)
+
+csapi = resultDict.get("CSAPI")
+if csapi and csapi.csModified:
+  if dryRun:
+    gLogger.notice("There are changes to Registry ready to commit, skipped because of dry run")
+  else:
+    yn = raw_input("There are changes to Registry ready to commit, do you want to proceed [default yes] [yes|no]:")
```
Comment (andresailer) [inline ConfigurationSystem/scripts/dirac-admin-voms-sync.py]: should be `[Y|n]` given the check in the next line (upper case marking the default choice). Though shouldn't the default be "no" ?

Comment (atsareg) [reply ConfigurationSystem/scripts/dirac-admin-voms-sync.py]: The check allows both Y and y and I will make it allowing also Yes and yes

**Comment by atsareg** [comment]
@fstagni: This is mostly a code refactoring to allow invoking in command line the agent logic. That is why the bulk of the code did not change. Refactoring the long method will be error prone I think. I will try to add some unit tests although the behavior depends on the external information which should be somehow mocked-up in some ugly way

**Comment by atsareg** [comment]
Current format test is failing because the pylint comment to a culprit line fails because the line was split by autopep8. I kind of remember similar problems before. Any clue how this can be solved ?

File: ConfigurationSystem/scripts/dirac-admin-voms-sync.py (hunk 4 above)
Comment (andresailer) [reply ConfigurationSystem/scripts/dirac-admin-voms-sync.py]: I knew that both Y and y were working, but some command line tools use capitalisation to indicate the default value, thus `[Y|n]`

Still my question is if the default here shouldn't be NO do no commit... so  `[y|N]`

**Comment by andresailer** [comment]
the format is failing because the lines are too long
```
./ConfigurationSystem/Agent/VOMS2CSAgent.py:82:121: E501 line too long (144 > 120 characters)
./ConfigurationSystem/Agent/VOMS2CSAgent.py:114:121: E501 line too long (152 > 120 characters)
```
Or do you mean how to split the line and still ignore the error?
```python
result = self.__syncCSWithVOMS(vomsSync,
                               proxyUserName=voAdminUser,  # pylint: disable=unexpected-keyword-arg
                               proxyUserGroup=voAdminGroup)  # pylint: disable=unexpected-keyword-arg
result = self.__addHomeDirectory(vo, newUsers,
                                 proxyUserName=voAdminUser,  # pylint: disable=unexpected-keyword-arg
                                 proxyUserGroup=voAdminGroup)  # pylint: disable=unexpected-keyword-arg
```

**Comment by atsareg** [comment]
Yes, I mean how to split the line and still ignore the error

File: ConfigurationSystem/scripts/dirac-admin-voms-sync.py (hunk 4 above)
Comment (atsareg) [reply ConfigurationSystem/scripts/dirac-admin-voms-sync.py]: This is the matter of taste, I think. If you do not make a dry run, then you most probably intend to commit the changes. So, I would prefer the default to be Yes.

File: ConfigurationSystem/Agent/VOMS2CSAgent.py [hunk 5]
```diff
@@ -6,33 +6,25 @@
...
-from DIRAC.Core.Utilities.List                         import fromChar
-from DIRAC.Core.Utilities.PrettyPrint                  import printTable
-from DIRAC.Resources.Catalog.FileCatalog               import FileCatalog
+from DIRAC.Core.Base.AgentModule import AgentModule
+from DIRAC.FrameworkSystem.Client.NotificationClient import NotificationClient
+from DIRAC import S_OK, gConfig
+from DIRAC.ConfigurationSystem.Client.Helpers.Registry import getVOOption, getUserOption
+from DIRAC.ConfigurationSystem.Client.VOMS2CSSynchronizer import VOMS2CSSynchronizer
+from DIRAC.Core.Utilities.Proxy import executeWithUserProxy
+from DIRAC.Resources.Catalog.FileCatalog import FileCatalog
```
Comment (fstagni) [inline ConfigurationSystem/Agent/VOMS2CSAgent.py]: [minor] imports are not grouped.

File: ConfigurationSystem/Agent/VOMS2CSAgent.py [hunk 6]
```diff
@@ -42,582 +34,145 @@ def __init__( self, *args, **kwargs ):
...
-        voAdminMail = getUserOption( voAdminUser, "Email")
-      voAdminGroup = getVOOption( vo, "VOAdminGroup", getVOOption( vo, "DefaultGroup" ) )
+        voAdminMail = getUserOption(voAdminUser, "Email")
+      voAdminGroup = getVOOption(vo, "VOAdminGroup", getVOOption(vo, "DefaultGroup"))
+
+      self.log.info('Performing VOMS sync for VO %s with credentials %s@%s' % (vo, voAdminUser, voAdminGroup))
+
+      autoAddUsers = getVOOption(vo, "AutoAddUsers", self.autoAddUsers)
+      autoModifyUsers = getVOOption(vo, "AutoModifyUsers", self.autoModifyUsers)
+      autoDeleteUsers = getVOOption(vo, "AutoDeleteUsers", self.autoDeleteUsers)
```
[the remark this answers is not in the document]
Comment (fstagni) [reply ConfigurationSystem/Agent/VOMS2CSAgent.py]: still.

File: ConfigurationSystem/Agent/VOMS2CSAgent.py [hunk 7]
```diff
@@ -42,582 +34,149 @@ def __init__( self, *args, **kwargs ):
...
-        else:
-          self.__adminMsgs[ 'Errors' ].append( 'Error in deleting users from CS:\n  %s' % str( oldUsers ) )
-          self.log.error( 'Error while user deletion from CS', result )
-      else:
-        self.__adminMsgs[ 'Info' ].append( 'The following users to be checked for deletion:\n  %s' % str( oldUsers ) )
-        self.log.info( 'The following users to be checked for deletion: %s' % str( oldUsers ) )
-
-    return S_OK( resultDict )
+  def __syncCSWithVOMS(self, vomsSync):
+    return vomsSync.syncCSWithVOMS()
```
Comment (fstagni) [inline ConfigurationSystem/Agent/VOMS2CSAgent.py]: Do we need this method at all?

File: ConfigurationSystem/Client/VOMS2CSSynchronizer.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+from collections import defaultdict
+
+from DIRAC.ConfigurationSystem.Client.CSAPI import CSAPI
+from DIRAC.Core.Security.VOMSService import VOMSService
+from DIRAC import S_OK, S_ERROR, gLogger, gConfig
+from DIRAC.ConfigurationSystem.Client.Helpers.Registry import getVOOption, getVOMSRoleGroupMapping, \
+    getUsersInVO, getAllUsers
+from DIRAC.Core.Utilities.List import fromChar
+from DIRAC.Core.Utilities.PrettyPrint import printTable
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSynchronizer.py]: [minor] imports are not grouped.

File: ConfigurationSystem/Client/VOMS2CSSynchronizer.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+        return result
+    result = self.csapi.describeUsers(diracUsers)
+    if not result['OK']:
+      self.log.error('Could not retrieve CS User description')
+    return result
+
+  def getVOUserReport(self):
+    """
+
+    :param str vo: VO name
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSynchronizer.py]: Not there.

File: ConfigurationSystem/Client/VOMS2CSSynchronizer.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+        self.log.info('The following users to be checked for deletion: %s' % str(oldUsers))
+
+    resultDict['CSAPI'] = self.csapi
+    resultDict['AdminMessages'] = self.adminMsgs
+    return S_OK(resultDict)
+
+  def getVOUserData(self, refreshFlag=False):
+    """ Get a report for users of a given VO
+
+    :param str vo: VO name
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSynchronizer.py]: Not there.

File: ConfigurationSystem/Client/VOMS2CSSynchronizer.py [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+    self.vomsVOName = getVOOption(vo, "VOMSName", "")
+    if not self.vomsVOName:
+      raise Exception("VOMS name not defined for VO %s" % vo)
+    self.adminMsgs = {'Errors': [], 'Info': []}
+    self.vomsUserDict = {}
+    self.autoModifyUsers = autoModifyUsers
+    self.autoAddUsers = autoAddUsers
+    self.autoDeleteUsers = autoDeleteUsers
+
+  def syncCSWithVOMS(self):
```
Comment (fstagni) [inline ConfigurationSystem/Client/VOMS2CSSynchronizer.py]: maybe a docstring?

**Comment by fstagni** [comment]
Andrei, please just fix at least the documentation and the docstring and then we merge.

File: ConfigurationSystem/Agent/VOMS2CSAgent.py (hunk 7 above)
Comment (atsareg) [reply ConfigurationSystem/Agent/VOMS2CSAgent.py]: This is a method wrapped in @executeWithUserProxy decorator

File: ConfigurationSystem/Agent/VOMS2CSAgent.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (atsareg) [reply ConfigurationSystem/Agent/VOMS2CSAgent.py]: These options can defined in the Agent configuration for all the VOs and then overriden per VO specifically

**Comment by fstagni** [comment]
Pylint spots:

```
************* Module DIRAC.WorkloadManagementSystem.private.ConfigHelper
WorkloadManagementSystem/private/ConfigHelper.py:34: [E0602(undefined-variable), findGenericPilotCredentials] Undefined variable 'pilotGroups'
WorkloadManagementSystem/private/ConfigHelper.py:35: [E0602(undefined-variable), findGenericPilotCredentials] Undefined variable 'pilotGroups'
```

**Pull request merged**